src/tests/Microsoft.PowerFx.Core.Tests/BinderTests.cs

```diff
@@ -0,0 +1,139 @@
+using Microsoft.PowerFx.Core.Texl;
+using Microsoft.PowerFx.Core.Texl.Builtins;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Tests;
+using Microsoft.PowerFx.Types;
+using Xunit;
+
+namespace Microsoft.PowerFx.Core.Tests
+{
+    public class BinderTests : PowerFxTest
+    {
+        [Theory]
+        [InlineData("x")]
+        [InlineData("x.Value")]
+        [InlineData("x.Tbl")]
+        [InlineData("First(x.Tbl)")]
+        [InlineData("First(x.Tbl).Value2")]
+        [InlineData("Last(x.Tbl).Value2")]
+        [InlineData("Index(x.Tbl, 3).Value2")]
+        [InlineData("x.Rec")]
+        [InlineData("x.Rec.Value3")]
+        [InlineData("tbl")]
+        [InlineData("First(tbl)")]
+        [InlineData("Last(tbl)")]
+        [InlineData("Index(tbl, 3)")]
+        public void TestMutableNodes(string expression)
+        {
+            var config = new PowerFxConfig();
+            config.SymbolTable.AddVariable(
+                "x",
+                new KnownRecordType(TestUtils.DT("![Value:n,Tbl:*[Value2:n],Rec:![Value3:n,Value4:n]]")),
+                mutable: true);
+            config.SymbolTable.AddVariable(
+                "tbl",
+                TableType.Empty().Add("Value", FormulaType.Number),
+                mutable: true);
+            var engine = new Engine(config);
+            var checkResult = engine.Check(expression);
+            Assert.True(checkResult.IsSuccess);
+            var binding = checkResult.Binding;
+            Assert.True(binding.IsMutable(binding.Top));
+        }
+
+        [Theory]
+        [InlineData("nf")]
+        [InlineData("const")]
+        [InlineData("FirstN(x.Tbl, 3).Value2")]
+        [InlineData("LastN(x.Tbl, 3).Value2")]
+        [InlineData("Filter(tbl, Value > 10)")]
+        [InlineData("FirstN(tbl, 2)")]
+        [InlineData("LastN(tbl, 2)")]
+        public void TestImmutableNodes(string expression)
+        {
+            var config = new PowerFxConfig();
+            config.SymbolTable.AddConstant("const", new NumberValue(IR.IRContext.NotInSource(FormulaType.Number), 2));
+            config.SymbolTable.AddVariable("nf", new TableType(TestUtils.DT("*[Value:n]")), mutable: false);
+            config.SymbolTable.AddVariable(
+                "x",
+                new KnownRecordType(TestUtils.DT("![Value:n,Tbl:*[Value2:n],Rec:![Value3:n,Value4:n]]")),
+                mutable: true);
+            config.SymbolTable.AddVariable(
+                "tbl",
+                TableType.Empty().Add("Value", FormulaType.Number),
+                mutable: true);
+            var engine = new Engine(config);
+            var checkResult = engine.Check(expression);
+            Assert.True(checkResult.IsSuccess);
+            var binding = checkResult.Binding;
+            Assert.False(binding.IsMutable(binding.Top));
+        }
+
+        private class FakePatchFunction : BuiltinFunction
```

Comment:
>FakePatchFunction [](http://example.com/codeflow?start=22&length=17)

What is significant about this override as opposed to other functions?

Reply:
Nothing really, just to showcase exactly what we want to block in the test below, without referencing the Patch function from this repository (since it's defined in MS.PowerFx.Interpreter, not here in Core). Once we move the *definition* of Patch from the interpreter to the core, we should be able to use that.